src/gwt/src/org/rstudio/studio/client/common/spelling/RealtimeSpellChecker.java

```diff
@@ -0,0 +1,313 @@
+      ValueChangeHandler<Boolean> prefChangedHandler = (event) -> context_.invalidateAllWords();
+      ValueChangeHandler<Boolean> realtimeChangedHandler = (event) -> {};
+      ValueChangeHandler<String> dictChangedHandler = (event) -> {};
+      userPrefs_.ignoreUppercaseWords().addValueChangeHandler(prefChangedHandler);
+      userPrefs_.ignoreWordsWithNumbers().addValueChangeHandler(prefChangedHandler);
+      userPrefs_.spellingDictionaryLanguage().addValueChangeHandler(dictChangedHandler);
+      userPrefs_.realTimeSpellchecking().addValueChangeHandler(realtimeChangedHandler);
+
+      // subscribe to user dictionary changes
+      context_.releaseOnDismiss(userDictionary_.addListChangedHandler((ListChangedEvent event) ->
+         {
+            // detect whether this is the first delivery of the list
+            // or if it is an update
+            userDictionaryWords_ = event.getList();
+            updateIgnoredWordsIndex();
+         }
+      ));
+   }
+
+   @Inject
+   void initialize(SpellingService spellingService, WorkbenchListManager workbenchListManager, UserPrefs uiPrefs)
+   {
+      spellingService_ = spellingService;
+      userDictionary_ = workbenchListManager.getUserDictionaryList();
+      userPrefs_ = uiPrefs;
+
+      if (domainSpecificWords_.isEmpty())
+      {
+         String[] words = RES.domainSpecificWords().getText().split("[\\r\\n]+");
+         for (String w : words)
+         {
+            if (w.length() > 0)
+               domainSpecificWords_.add(w.toLowerCase());
+         }
+      }
+   }
+   
+   public boolean realtimeSpellcheckEnabled()
+   {
+      return userPrefs_.realTimeSpellchecking().getValue();
+   }
+
+   public void addToUserDictionary(final String word)
+   {
+      // append to user dictionary (this is async so the in-memory list of
+      // ignored words won't update immediately)
+      userDictionary_.append(word);
+      
+      // add the words to the ignore list (necessary for contexts that 
+      // rely on the word being on the ignore list for correct handling)
+      // (note that allIgnoredWords_ will soon be overwritten after the
+      // userDictionary list changed handler is invoked)
+      allIgnoredWords_.add(word);
+      
+      // invalidate the context
+      context_.invalidateWord(word, true);
+   }
+   
+   public boolean isIgnoredWord(String word)
+   {
+      return contextDictionary_.contains(word);
+   }
+
+   public void addIgnoredWord(String word)
+   {
+      contextDictionary_.add(word);
+      writeContextDictionary(word);
+   }
+   
+   public void removeIgnoredWord(String word)
+   {
+      contextDictionary_.remove(word);
+      writeContextDictionary(word);
+   }
+   
+   private void writeContextDictionary(String affectedWord)
+   {
+      context_.writeDictionary(contextDictionary_);
+      updateIgnoredWordsIndex();
+      context_.invalidateWord(affectedWord, false);
+   }
+
+   public SpellCheckerResult cachedWords(ArrayList<String> words)
```

Comment:
Nit: this function doesn't really suggest that it would mutate the passed-in argument `words`, but it looks like it does? IIUC the intention is to separate words for which we have some locally-cached result, versus those that require an RPC for the spellcheck operation?

Reply:
I'd suggest either a more evocative name (e.g. `checkAndRemoveCachedWords()`) or perhaps an "output" parameter, e.g. `uncachedWords`, that gets filled when this function is called.

Reply:
Good catch. The removal is actually just leftover cruft from before I refactored the function out. I'll rename the function to `getCachedWords` and remove the removes. It's not intended to mutate at all.